python-oop/intro-to-design-patterns.md

```diff
@@ -0,0 +1,116 @@
+| Observer       | A class whose responsibility is to watch for changes over many objects, and notify many objects when changes happen. | Instead of many objects needing to refer to each other to announce changes, many objects can depend on one observer.                                                                                                                                                       |
+| Decorator      | A class or method whose responsibility is to dynamically alter and extend the behavior of a function or class.       | If many objects must be extended in the same way, one decorator can decorate all of them without changing their implementation.                                                                                                                                            |
+| Strategy       | A class whose responsibility is to represent an algorithm.                                                           | All relevant classes can use the strategy class instead of implementing the algorithm. If an algorithm needs to be updated, only the strategy class needs to change.                                                                                                       |
+
+## Overdesign or Wrong Design Leads to Unmaintainable Code
+
+Every design pattern has its pros and cons.
+
+### !callout-info
+
+## Back to Tatiana's Moving Example
+
+When the boxes are unloaded, Tatiana wishes that each box went directly to the correct room. Tatiana might accomplish this by having really good and clear labels on her boxes. However, that requires Tatiana to know what rooms exist and decide what room each box belongs to while she's packing. How Tatiana labels her boxes can make unloading easier (pro), but means she's responsible for knowing and doing more while packing (con).
+
+### !end-callout
+
+There is such a thing as "overdesigning," or using a design pattern that increases complexity way more than necessary.
+
+In general, it's often best to implement a solution without a design pattern, make a git commit, and then refactor.
+
+## Moving Forward: How Do We Use These?
+
+Design patterns are applied when it makes problem-solving easier. In general, it takes time, experience, research, context, and experimentation to apply design patterns effectively.
+
+When considering a design pattern, the team should consider:
+
+- Does this make the code more or less maintainable over time?
+- What is the cost of implementing this design pattern?
+- What are the benefits of implementing this design pattern?
+- What are the alternatives?
+
+## Check for Understanding
+
+<!-- Something like matching between design pattern and definition. x3 times. Test Iterator, Decorator, and Factory, as these are the most common patterns for a junior dev (IMHO) -->
```

Comment:
```suggestion
<!-- Something like matching between design pattern and definition. x3 times. Test Iterator, Decorator, and Factory, as these are the most common patterns for a junior dev (IMHO) -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: be55b24a-8a6e-4bcc-9661-f5354cdd59b6
* title: Intro to Design Patterns I

##### !question
Which of these options are design patterns?
##### !end-question

##### !options
* Iterator
* Function
* Decorator
* Factory 
* Adapter
##### !end-options

##### !answer
* Iterator
* Decorator
* Factory 
* Adapter
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: fae4bb8a-4879-44bc-83cb-3a3efb031908
* title: Intro to Design Patterns II

##### !question
Which options describes the design pattern of an Iterator?
##### !end-question

##### !options
* A class that can iterate through an object. Available to other classes to use, however, if the object changes, the iterator will need to change. 
* A while or for loop.
* A class whose responsibility is to know how to iterate through an object.
##### !end-options

##### !answer
* A class whose responsibility is to know how to iterate through an object.
* A class that can iterate through an object. Available to other classes to use, however, if the object changes, the iterator will need to change. 
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: fae4bb8a-4879-44bc-83cb-3a3efb031908
* title: Intro to Design Patterns III

##### !question
Which options describes the design pattern of a Factory?
##### !end-question

##### !options
* A class that can iterate through an object. Available to other classes to use, however, if the object changes, the iterator will need to change. 
* A class or method whose responsibility is to create instances of some class.
* A class that holds logic to create instances, which requires close attention to a constructor. If a constructor is updated, only the factory needs to change, rather than each line that calls the constructor. 
##### !end-options

##### !answer
* A class or method whose responsibility is to create instances of some class.
* A class that holds logic to create instances, which requires close attention to a constructor. If a constructor is updated, only the factory needs to change, rather than each line that calls the constructor. 
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: b64da97d-08cb-48d9-9165-2010c70bdcca
* title: Intro to Design Patterns IV

##### !question
Which options describes the design pattern of a Decorator?
##### !end-question

##### !options
* A class or method whose responsibility is to dynamically alter and extend the behavior of a function or class.  
* A class or method whose responsibility is to decorate a project with syntactical sugar.
* A class or method that can extend multiple objects without changing their implementation.
##### !end-options

##### !answer
* A class or method whose responsibility is to dynamically alter and extend the behavior of a function or class.  
* A class or method that can extend multiple objects without changing their implementation.
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->
```